### DISCUSSION TITLE
array transfer between federates

### DISCUSSION THREAD

**Discussion by ajayornl** [GMLC-TDC/HELICS#2388] [General]
Hi, 

I am using the fundamental_default example (HELICS-Examples/user_guide_examples/fundamental/fundamental_default/) in which the battery and charger federates exchange current and voltage information. I am trying to modify this example for some other application. 

Herein, is it possible to exchange arrays or matrices (with complex elements) between federates. Moreover, is it possible to exchange something like structure array (https://www.mathworks.com/help/matlab/ref/struct.html) between federates. 

Thank you. 

Ajay

**Comment by phlptp** [comment]
HELICS directly supports vector transfer and complex vector transfer.    As far as structures go we have at various time looked at supporting it but there is no common standard for that.  And when done in different application the most common way is to convert it to json and send the string, so there hasn't been a strong case to support it directly.  

my suggestion would be to look at something like https://www.mathworks.com/help/matlab/ref/jsonencode.html for Matlab for sending more complex structured data.

**Comment by ajayornl** [comment]
Thanks for the response. Do you have any example in which you are exchanging vectors between two federates? 

I tried using the helicsPublicationPublishVector command, where I am hoping to send a fixed array between the battery and charger federates (in the default fundamental example). This gives an error. While using helicsPublicationPublishVector command, do we have to edit the json file as well?

**Comment by phlptp** [comment]
The type of the publication you set in the JSON file is the type of data that gets transferred.  HELICS does do conversion between all the supported types but generally a good idea to match up the specified type with the type of data you expect to send most of the time.  

What error are you getting.  I know that works in C/C++ what I don't know is whether the python interface is expecting a particular kind of array since I don't work in python all that often.  @kdheepak  can give a better explanation and help debug what is going on in your case.

**Comment by ajayornl** [comment]
Hi Philip, 
Thanks for the info. I will check with Dheepak regarding this. Btw, I see the following error, 


File "C:\Users\aaa\Anaconda3\envs\env01\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Users\aaa\Anaconda3\envs\env01\lib\site-packages\helics_cli\status_checker.py", line 40, in run
    raise HELICSRuntimeError("Error has occurred")
helics_cli.exceptions.HELICSRuntimeError: Error has occurred

**Comment by kdheepak** [comment]
If you are using Python, the simplest way is to convert the complex matrix to a complex array using reshape, and then take a "view" of it using float32 or float64. See example code below:

![Screen Shot 2022-07-05 at 1 28 00 PM](https://user-images.githubusercontent.com/1813121/177401342-af61c748-b23a-40e3-a82a-c9dfbbf83315.png)

You can then send this float64 array using `helicsPublicationPublishVector(...)`. You'll have to convert it back appropriately on the other side, but using numpy it should be straightforward to do that. 

There are similar functions available in MATLAB and Julia.

**Reply by ajayornl** [reply]
thanks for response Dheepak. Actually, I am getting error with sending a simple array such as np.array([1,2,3]) using helicsPublicationPublishVector(..). I have a working example in which a (sensor) federate is sending a scalar to another (estimator) federate. When I replace helicsPublicationPublishDouble with helicsPublicationPublishVector, I am getting error. Does the json file also needs to be edited? Following are my jsons: 

{
  "name": "Sensor",
  "core_type": "zmq",
  "log_level": "warning",
  "period": 10,
  "uninterruptible": false,
  "terminate_on_error": true,
  "wait_for_current_time_update": true,
  "publications": [
    {
      "global": true,
      "key": "Sensor/EV1_current",
      "type": "",
      "unit": ""
    }
  ]
}


{
  "core_name": "Estimator",
  "core_type": "zmq",
  "log_level": "warning",
  "name": "Charger",
  "period": 10,
  "subscriptions": [
    {
      "global": true,
      "key": "Sensor/EV1_current",
      "type": "",
      "unit": ""
    }
  ],
  "terminate_on_error": true,
  "uninterruptible": false
}

**Reply by kdheepak** [reply]
What is the error you are getting?

**Reply by kdheepak** [reply]
It is possible that `helicsPublicationPublishVector` doesn't work with numpy and only works with list of floats. Can you try:

```
h.helicsPublicationPublishVector(pub, [x for x in arr])
```

instead?

**Comment by ajayornl** [comment]
I publish using the command: h.helicsPublicationPublishVector(pubid, charging_current)

Following is the error: 

helics-cli [error]: Process Estimator exited with return code 1
helics-cli [error]: Process Sensor exited with return code 1
helics-cli [info]: Done.
Exception in thread Thread-1:
Traceback (most recent call last):
  File "C:\Users\ay\Anaconda3\envs\env01\lib\threading.py", line 932, in _bootstrap_inner
    self.run()
  File "C:\Users\ay\Anaconda3\envs\env01\lib\site-packages\helics_cli\status_checker.py", line 40, in run
    raise HELICSRuntimeError("Error has occurred")
helics_cli.exceptions.HELICSRuntimeError: Error has occurred

**Reply by kdheepak** [reply]
Can you share the full output of helics-cli? It should say something has exited with return code that is _not_ 1. There should also be a .log file file for that process in the folder when the runner.json file is located.

**Comment by ajayornl** [comment]
FYI, the code h.helicsPublicationPublishVector(pubid, [1,2,3,4]) is not giving any error

**Reply by kdheepak** [reply]
I'll look into what it would take to make numpy work but I think for now you'll have to convert to lists manually.
